File: share/webapp/js/room/receipt.js
/*
// Matrix Construct
//
// Copyright (C) Matrix Construct Developers, Authors & Contributors
// Copyright (C) 2016-2018 Jason Volk <jason@zemos.net>
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice is present in all copies. The
// full license for this software is available in the LICENSE file.
*/

'use strict';


/**************************************
 * Receipts
 *
 */

room.receipt = {};

room.receipt.__proto__ = function(room)
{
	if(!maybe(() => room.local.receipt))
		room.local.receipt = {};

	this.local = room.local.receipt;
};

room.receipt.send = async function(event_id, type = "m.read", opts = {})
{
	return false;

	if(this.control.receipt_in_progress)
		return false;

	this.control.receipt_in_progress = event_id;
	let request = client.m.rooms.receipt.post(this.id, event_id, type, opts); try
	{
		let data = await request.response;
		return;
	}
	finally
	{
		this.control.receipt_in_progress = null;
	}
};

room.receipt.send.current = function()
{
	let current = this.timeline.current_event();
	if(!current)
		return false;

	if(!current.event_id)
		return false;

	if(mc.event.is_local(current))
		return false;

	if(current.type == "m.receipt")
		return false;

	if(maybe(() => current["m.read"][mc.my.mxid]))
		return false;

	if(maybe(() => current.origin_server_ts <= this.local.ts))
		return false;

	let receipts =
	{
		[current.event_id]:
		{
			"m.read":
			{
				[mc.my.mxid]:
				{
					ts: mc.date.now(),
				},
			},
		},
	};

	this.local.ts = mc.now();
	this.receipt.update(receipts);
	return this.receipt.send(current.event_id);
};

room.receipt.update = function(receipts)
{
	Object.each(receipts, (event_id, read) =>
	{
		mayif(() => this.timeline.get(event_id), (event) =>
		{
			this.receipt.update.event(event, read);
		});
	});
};

/**
 * merges updates to an object now attached to the event "m.read"
 * which explicitly specified so other keys in the receipt may
 * not clobber.
 */
room.receipt.update.event = function(event, read)
{
	Object.update(event,
	{
		"m.read": read["m.read"],
	});
};

/**
 * Sorts the receipts by timestamp so they can be viewed in the order of
 * reception.
 *
 * THIS FUNCTION MUST HAVE STABLE RESULTS for Angular and receipts with
 * the same timestamp must have some alternative ordering property.
 *
 * @returns Array of mxids which are keys for receipts ordered by ts
 */
room.receipt.sorted = function(receipts)
{
	return Object.keys(receipts).sort((a, b) =>
	{
		let tsa = receipts[a].ts;
		let tsb = receipts[b].ts;

		if(tsa === undefined && tsb === undefined)
			return 0;
		else if(tsa === undefined)
			return -1;
		else if(tsb === undefined)
			return 1;
		else
			return tsa < tsb?  -1:
			       tsa > tsb?   1:
			       a < b?      -1:
			       a > b?       1:
			                    0;
	});

	return ret;
};

File: modules/client/sync/rooms/receipt.cc
// Matrix Construct
//
// Copyright (C) Matrix Construct Developers, Authors & Contributors
// Copyright (C) 2016-2018 Jason Volk <jason@zemos.net>
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice is present in all copies. The
// full license for this software is available in the LICENSE file.

ircd::mapi::header
IRCD_MODULE
{
    "Client Sync :Room Receipts"
};

namespace ircd::m::sync
{
	static void _reformat_receipt(json::stack::object &, const m::event &);
	static void _handle_receipt(data &, const m::event &);
	static void _handle_user(data &, const m::user &);
	static void room_ephemeral_m_receipt_m_read_polylog(data &);
	extern item room_ephemeral_m_receipt_m_read;
}

decltype(ircd::m::sync::room_ephemeral_m_receipt_m_read)
ircd::m::sync::room_ephemeral_m_receipt_m_read
{
	"rooms.ephemeral.m_receipt",
	room_ephemeral_m_receipt_m_read_polylog
};

void
ircd::m::sync::room_ephemeral_m_receipt_m_read_polylog(data &data)
{
	const m::room &room{*data.room};
	const m::room::members members
	{
		room
	};

	members.for_each(data.membership, m::room::members::closure{[&data]
	(const m::user::id &user_id)
	{
		_handle_user(data, user_id);
	}});
}

void
ircd::m::sync::_handle_user(data &data,
                            const m::user &user)
{
	static const m::event::fetch::opts fopts
	{
		m::event::keys::include
		{
			"event_id", "content", "sender",
		},
	};

	m::user::room user_room{user};
	user_room.fopts = &fopts;
	if(head_idx(std::nothrow, user_room) < data.range.first)
		return;

	const m::room::id &room_id{*data.room};
	user_room.get(std::nothrow, "ircd.read", room_id, [&data]
	(const m::event &event)
	{
		if(apropos(data, event))
			_handle_receipt(data, event);
	});
}

void
ircd::m::sync::_handle_receipt(data &data,
                               const m::event &event)
{
	const json::object content
	{
		at<"content"_>(event)
	};

	data.commit();
	json::stack::object object
	{
		data.out
	};

	// type
	json::stack::member
	{
		object, "type", "m.receipt"
	};

	// content
	json::stack::object content_
	{
		object, "content"
	};

	json::stack::object event_id_
	{
		content_, unquote(content.at("event_id"))
	};

	json::stack::object m_read_
	{
		event_id_, "m.read"
	};

	json::stack::object sender_
	{
		m_read_, at<"sender"_>(event)
	};

	json::stack::member
	{
		sender_, "ts", json::value(content.at("ts"))
	};
}
